R/flatten.R

```diff
@@ -0,0 +1,118 @@
+#' Flatten `dm` into a wide table
+#'
+#' This function joins all the tables of your `dm` object together and returns one
+#' table containing all unique columns. This is especially useful in connection
+#' with `cdm_select_tbl()` in order to reduce the number of tables in advance.
+#'
+#' @inheritParams cdm_join_tbl
+#' @param sep In case there are columns with non-unique names, they will be renamed using
+#' [`cdm_disambiguate_cols()`]. `sep` is the character variable separating the table
+#' names and the ambiguous column names
+#' @family Flattening functions
+#'
+#' @return A wide table resulting of consecutively joining all tables together.
+#'
+#' @examples
+#' cdm_nycflights13() %>%
+#' cdm_select_tbl(-weather) %>%
+#' cdm_flatten()
+#'
+#' @export
+cdm_flatten <- function(dm, join = left_join, sep = ".") {
+  if (!are_all_vertices_connected(create_graph_from_dm(dm), src_tbls(dm))) {
+    abort_vertices_not_connected("cdm_flatten")
+  }
+  clean_dm <- cdm_apply_filters(dm) %>%
+      cdm_disambiguate_cols(sep = sep, quiet = TRUE) %>%
+      adapt_fk_cols()
+  g <- create_graph_from_dm(clean_dm, directed = TRUE)
+  filtered_tables <- clean_dm %>% cdm_get_tables()
+
+  # chose a "child table" (only outgoing FKs) as a start (otherwise we might
+  # end up with empty rows, if "parent/dim table" contains PK values that are
+  # not present in child table (we might anyway though, depending on `join`-type))
+  initial_LHS_name <- g %>%
+    igraph::topo_sort() %>%
+    names() %>%
+    pluck(1)
+  initial_LHS <- filtered_tables[[initial_LHS_name]]
+
+  # each next table needs to be accessible from the former table; this is not ensured by `topo_sort()`
+  # we achieve this with a breadth-first-search (BFS)
+  order <- igraph::bfs(g, initial_LHS_name) %>%
+    extract2("order") %>%
+    names()
+
+  # early return in case of only one table...
+  if (length(order) == 1) return(initial_LHS)
+  # since we want to work with `reduce2()`, the `.init`-parameter is the first table
+  # in `order` and the next table is always the join partner
+  ordered_table_list <- filtered_tables[order] %>%
```

Comment:
Do you want the line `if (length(order) == 1) return(initial_LHS)` to be removed? (just asking, because if it stays the code below doesn't matter anyway in case of just one table)

Reply:
I prefer to avoid redundant code where possible. I'm fine with a comment that indicates that the operations afterwards are a no-op in this case.

Reply:
I'm rather certain that latest when the joins are to be made an error occurs if a join should be made between a table and an empty list. Otherwise put, I am not sure how to avoid the early return line.